sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -0,0 +1,171 @@
+
+  def resolveVariable(e: Expression) : Expression = {
+    /* We know that the expression is either UnresolvedAttribute or Alias,
+     * as passed from the parser.
+     * If it is an UnresolvedAttribute, we look it up in the catalog and return it.
+     * If it is an Alias, we resolve the child and return an Alias with the same name.
+     */
+    e match {
+      case u: UnresolvedAttribute =>
+        lookupVariable(u.nameParts) match {
+          case Some(variable) =>
+            variable.copy(canFold = false)
+          case _ => throw unresolvedVariableError(u.nameParts, Seq("SYSTEM", "SESSION"))
+        }
+
+      case a: Alias =>
+        Alias(resolveVariable(a.child), a.name)()
+
+      case other => throw SparkException.internalError(
+        "Unexpected variable expression in ParametrizedQuery: " + other)
+    }
+  }
+
+  def resolveArguments(expressions : Seq[Expression]): Seq[Expression] = {
+    expressions.map { exp =>
+      if (exp.resolved) {
+        exp
+      } else {
+        resolveVariable(exp)
+      }
+    }
+  }
+
+  def extractQueryString(either : Either[String, UnresolvedAttribute]) : String = {
+    either match {
+        case Left(v) => v
+        case Right(u) =>
+          val varReference = lookupVariable(u.nameParts) match {
+            case Some(variable) => variable.copy(canFold = false)
+            case _ => throw unresolvedVariableError(u.nameParts, Seq("SYSTEM", "SESSION"))
+          }
+
+          if (!varReference.dataType.sameType(StringType)) {
+            throw new AnalysisException(
+              errorClass = "INVALID_VARIABLE_TYPE_FOR_QUERY_EXECUTE_IMMEDIATE",
+              messageParameters = Map(
+                "varType" -> varReference.dataType.simpleString))
+          }
+
+          // call eval with null row.
+          // this is ok as this is variable and invoking eval should
+          // be independent of row
+          varReference.eval(null).toString
+      }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(EXECUTE_IMMEDIATE), ruleId) {
+    case ExecuteImmediateQuery(expressions, query, targetVariablesOpt) =>
+      val queryString = extractQueryString(query)
+      val plan = CatalystSqlParser.parsePlan(queryString);
+
+      val posNodes = plan.collect {
+        case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: PosParameter => n })
+      }.flatten
+      val namedNodes = plan.collect {
+        case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect{ case n: NamedParameter => n })
+      }.flatten
+
+      val queryPlan = if (expressions.isEmpty || (posNodes.isEmpty && namedNodes.isEmpty)) {
+        plan
+      } else if (posNodes.nonEmpty && namedNodes.nonEmpty) {
+        throw new AnalysisException(
+          errorClass = "INVALID_QUERY_BOTH_POSITIONAL_AND_NAMED_PARAMETERS_PRESENT",
+          messageParameters = Map.empty)
+      } else {
```

Comment:
you can drop the `else` since you raise an error in the previous case, de-denting the rest of the function.